cmd/zed/agents/zfs_mod.c

```diff
@@ -0,0 +1,911 @@
+ *
+ */
+
+#include <ctype.h>
+#include <devid.h>
+#include <fcntl.h>
+#include <libnvpair.h>
+#include <libzfs.h>
+#include <limits.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+#include <syslog.h>
+#include <sys/list.h>
+#include <sys/sunddi.h>
+#include <sys/sysevent/eventdefs.h>
+#include <sys/sysevent/dev.h>
+#include <pthread.h>
+#include <unistd.h>
+#include "zfs_agents.h"
+#include "../zed_log.h"
+
+#define	DEV_BYID_PATH	"/dev/disk/by-id/"
+#define	DEV_BYPATH_PATH	"/dev/disk/by-path/"
+
+typedef void (*zfs_process_func_t)(zpool_handle_t *, nvlist_t *, boolean_t);
+
+libzfs_handle_t *g_zfshdl;
+list_t g_pool_list;	/* list of unavailable pools at initialization */
+list_t g_device_list;	/* list of disks with asynchronous label request */
+boolean_t g_enumeration_done;
+pthread_t g_zfs_tid;
+
+typedef struct unavailpool {
+	zpool_handle_t	*uap_zhp;
+	pthread_t	uap_enable_tid;	/* dataset enable thread if activated */
+	list_node_t	uap_node;
+} unavailpool_t;
+
+typedef struct pendingdev {
+	char		pd_physpath[128];
+	list_node_t	pd_node;
+} pendingdev_t;
+
+static int
+zfs_toplevel_state(zpool_handle_t *zhp)
+{
+	nvlist_t *nvroot;
+	vdev_stat_t *vs;
+	unsigned int c;
+
+	verify(nvlist_lookup_nvlist(zpool_get_config(zhp, NULL),
+	    ZPOOL_CONFIG_VDEV_TREE, &nvroot) == 0);
+	verify(nvlist_lookup_uint64_array(nvroot, ZPOOL_CONFIG_VDEV_STATS,
+	    (uint64_t **)&vs, &c) == 0);
+	return (vs->vs_state);
+}
+
+static int
+zfs_unavail_pool(zpool_handle_t *zhp, void *data)
+{
+	zed_log_msg(LOG_INFO, "zfs_unavail_pool: examining '%s' (state %d)",
+	    zpool_get_name(zhp), (int)zfs_toplevel_state(zhp));
+
+	if (zfs_toplevel_state(zhp) < VDEV_STATE_DEGRADED) {
+		unavailpool_t *uap;
+		uap = malloc(sizeof (unavailpool_t));
+		uap->uap_zhp = zhp;
+		uap->uap_enable_tid = 0;
+		list_insert_tail((list_t *)data, uap);
+	} else {
+		zpool_close(zhp);
+	}
+	return (0);
+}
+
+/*
+ * Remove partition suffix from a vdev path.  Partition suffixes may take three
+ * forms: "-partX", "pX", or "X", where X is a string of digits.  The second
+ * case only occurs when the suffix is preceded by a digit, i.e. "md0p0" The
+ * third case only occurs when preceded by a string matching the regular
+ * expression "^([hsv]|xv)d[a-z]+", i.e. a scsi, ide, virtio or xen disk.
+ */
+static void
+zfs_strip_partition(char *path)
+{
+	char *part = NULL, *d = NULL;
+
+	path = strrchr(path, '/');
+	path++;
+
+	if ((part = strstr(path, "-part")) && part != path) {
+		d = part + 5;
+	} else if ((part = strrchr(path, 'p')) &&
+	    part > path + 1 && isdigit(*(part-1))) {
+		d = part + 1;
+	} else if ((path[0] == 'h' || path[0] == 's' || path[0] == 'v') &&
+	    path[1] == 'd') {
+		for (d = &path[2]; isalpha(*d); part = ++d);
+	} else if (strncmp("xvd", path, 3) == 0) {
+		for (d = &path[3]; isalpha(*d); part = ++d);
+	}
+	if (part && d && *d != '\0') {
+		for (; isdigit(*d); d++);
+		if (*d == '\0')
+			*part = '\0';
+	}
+}
```

Comment:
Why not use the existing `strip_partition` function in libzfs.so?  The zed already links against the library and it would be easy to add to the public header file.